votor/src/event_handler/stats.rs

```diff
@@ -0,0 +1,334 @@
+        entry.count = entry.count.saturating_add(1);
+        entry.time_us = entry.time_us.saturating_add(time_us);
+    }
+
+    pub fn incr_vote(&mut self, bls_op: &BLSOp) {
+        if let BLSOp::PushVote { message, .. } = bls_op {
+            let ConsensusMessage::Vote(vote) = **message else {
+                warn!("Unexpected BLS message type: {message:?}");
+                return;
+            };
+            let vote_type = VoteType::get_type(&vote.vote);
+            let entry = self.sent_votes.entry(vote_type).or_insert(0);
+            *entry = entry.saturating_add(1);
+            if vote_type == VoteType::Notarize {
+                let entry = self.slot_tracking_map.entry(vote.vote.slot()).or_default();
+                entry.vote_notarize = Some(Instant::now());
+            } else if vote_type == VoteType::Skip {
+                let entry = self.slot_tracking_map.entry(vote.vote.slot()).or_default();
+                entry.vote_skip = Some(Instant::now());
+            }
+        } else {
+            warn!("Unexpected BLS operation: {bls_op:?}");
+        }
+    }
+
+    pub fn maybe_report(&mut self) {
+        let now = Instant::now();
+        if now.duration_since(self.last_report_time) < STATS_REPORT_INTERVAL {
+            return;
+        }
+        datapoint_info!(
+            "event_handler_stats",
+            ("ignored", self.ignored as i64, i64),
+            (
+                "leader_window_replaced",
+                self.leader_window_replaced as i64,
+                i64
+            ),
+            ("set_root_count", self.set_root_count as i64, i64),
+            ("timeout_set", self.timeout_set as i64, i64),
+        );
+        for (event, EventCountAndTime { count, time_us }) in &self.received_events_count_and_timing
+        {
+            datapoint_info!(
+                "event_handler_received_event_count_and_timing",
+                ("event", format!("{:?}", event), String),
+                ("count", *count as i64, i64),
+                ("elapsed_us", *time_us as i64, i64)
+            );
+        }
+        datapoint_info!(
+            "event_handler_timing",
+            (
+                "receive_event_time_us",
+                self.receive_event_time_us as i64,
+                i64
+            ),
+            (
+                "send_votes_batch_time_us",
+                self.send_votes_batch_time_us as i64,
+                i64
+            ),
+        );
+        for (vote_type, count) in &self.sent_votes {
+            datapoint_info!(
+                "event_handler_sent_vote_count",
+                ("vote", format!("{:?}", vote_type), String),
+                ("count", *count as i64, i64)
+            );
+        }
+
+        // Only report slots lower than the `root_slot`
+        let split_off_map = self.slot_tracking_map.split_off(&self.root_slot);
+        for (slot, tracking) in &self.slot_tracking_map {
+            tracking.report(*slot);
+        }
```

Comment:
I'm debating whether we should move this outside of the logic that runs every second. Namely, should we report it immediately when we know we have a new root ? I guess maybe not necessary since we record the finalization in `SlotTracking`

Reply:
Works for me to handle it later